src/vs/workbench/contrib/notebook/browser/notebookExecutionStateServiceImpl.ts

```diff
@@ -140,51 +140,68 @@ export class NotebookExecutionStateService extends Disposable implements INotebo
 
-		notebook.onWillAddRemoveCells((e: NotebookTextModelWillAddRemoveEvent) => {
-			const lastFailedCell = this.getLastFailedCellForNotebook(notebook.uri);
-			e.rawEvent.changes.forEach(([start, deleteCount]) => {
-				if (deleteCount) {
-					const deletedHandles = notebook.cells.slice(start, start + deleteCount).map(c => c.handle);
-					if (lastFailedCell && deletedHandles.includes(lastFailedCell)) {
-						this._setLastFailedCellVisibility(notebook.uri, false);
-					}
-				}
-			});
-		});
+		return exe;
+	}
 
-		notebook.onDidChangeContent((events) => {
-			const lastFailedCell = this.getLastFailedCellForNotebook(notebook.uri);
-			events.rawEvents.forEach((e) => {
-				if (e.kind === NotebookCellsChangeType.ModelChange) {
-					e.changes.forEach(([start, deleteCount]) => {
-						if (deleteCount === 0) {
-							const addHandle = notebook.cells[start].handle;
-							if (addHandle === lastFailedCell) {
-								this._setLastFailedCellVisibility(notebook.uri, true);
-							}
-						}
-					});
-				}
-			});
-		});
+	private _setLastFailedCell(notebookURI: URI, cellHandle: number) {
+		this._lastFailedCells.set(notebookURI, cellHandle);
 
-		return exe;
+		if (this._lastFailedCellDisposableListeners.get(notebookURI) === undefined) {
+			const notebook = this._notebookService.getNotebookTextModel(notebookURI);
+			if (notebook) {
+				this._setFailedCellListeners(notebook);
+			}
+		}
+		this._onDidChangeLastRunFailState.fire({ failed: true, notebook: notebookURI });
 	}
 
-	private _setLastFailedCellVisibility(notebook: URI, visible: boolean) {
-		this._lastFailedCellsVisible.set(notebook, visible);
-		this._onDidChangeLastRunFailState.fire({ failed: visible, notebook });
+	private _clearLastFailedCell(notebookURI: URI) {
+		this._lastFailedCells.delete(notebookURI);
+		this._removeFailedCellListeners(notebookURI);
+		this._onDidChangeLastRunFailState.fire({ failed: false, notebook: notebookURI });
 	}
 
-	private _setLastFailedCell(notebook: URI, cellHandle: number) {
-		this._lastFailedCells.set(notebook, cellHandle);
-		this._lastFailedCellsVisible.set(notebook, true);
-		this._onDidChangeLastRunFailState.fire({ failed: true, notebook });
+	private _setFailedCellListeners(notebook: NotebookTextModel) {
+		const lastFailedCellDisposableListener = combinedDisposable(
+			notebook.onWillAddRemoveCells((e: NotebookTextModelWillAddRemoveEvent) => {
+				const lastFailedCell = this.getLastFailedCellForNotebook(notebook.uri);
+				e.rawEvent.changes.some(([start, deleteCount]) => {
+					if (deleteCount) {
+						const lastFailedCellPos = notebook.cells.findIndex(c => c.handle === lastFailedCell);
+						if (lastFailedCellPos >= start && lastFailedCellPos < start + deleteCount) {
+							this._onDidChangeLastRunFailState.fire({ failed: false, notebook: notebook.uri });
+							return true;
+						}
+					}
+					return false;
+				});
+			}),
+			notebook.onDidChangeContent((events) => {
```

Comment:
Maybe I led you to the wrong event, do you need this one too? When is this fired?

Reply:
Is `onWillAddRemoveCells` not fired for undo/redo?

Reply:
hmm, I thought that one of them was returning an off-by-one on the cell handle when re-doing but I just checked and it seems fine; will fix

Reply:
Tried implementing both listening for add/remove on `onWillAddRemoveCells` and I get an issue where the listener for the re-add (pressing 'undo' on delete) is always off by one in the handler of the added cell. Might have to do with each of them get called.

Reply:
ok I looked more info it. seems that `onWillAddRemoveCells` will run before the add/delete and `onDidChangeContent` will run after (ie: https://github.com/microsoft/vscode/blob/559e9beea981b47ffd76d90158ccccafef663324/src/vs/workbench/contrib/notebook/common/model/notebookTextModel.ts#L734). Therefore, I catch deletes using `onWillAddRemoveCells` because I need the list of cells before the deletion and `onDidChangeContent` for catching re-adds because i need the state after the re-add

Reply:
> because i need the state after the re-add

If you look at the list of new items on the event, here, https://github.com/microsoft/vscode/blob/a951adfd4facdc9244c5608a159d9c7b7cd6b95e/src/vs/workbench/contrib/notebook/common/notebookCommon.ts#L244

can you look at that to use `onWillAddRemoveCells` for both?